go/felix/ifacemonitor/iface_monitor.go

```diff
@@ -89,26 +106,100 @@ func (m *InterfaceMonitor) handleNetlinkUpdate(update netlink.LinkUpdate) {
+	addr := update.LinkAddress.IP.String()
+	ifIndex := update.LinkIndex
+	exists := update.NewAddr
+	log.WithFields(log.Fields{
+		"addr":    addr,
+		"ifIndex": ifIndex,
+		"exists":  exists,
+	}).Info("Netlink address update.")
+
+	if _, known := m.ifaceName[ifIndex]; !known {
+		log.WithField("ifIndex", ifIndex).Warn("No known iface with this index.")
+		return
+	}
+	if _, known := m.ifaceAddrs[ifIndex]; !known {
+		// We think this interface is down, or haven't yet heard about it at all -
+		// indicates a race between the link and address update channels.
+		// Addresses will be notified when we process the link update.
+		log.WithField("ifIndex", ifIndex).Warn("Race for new interface.")
+		return
+	}
+
+	if exists {
+		if !m.ifaceAddrs[ifIndex].Contains(addr) {
+			m.ifaceAddrs[ifIndex].Add(addr)
+			m.notifyIfaceAddrs(ifIndex)
+		}
+	} else {
+		if m.ifaceAddrs[ifIndex].Contains(addr) {
+			m.ifaceAddrs[ifIndex].Discard(addr)
+			m.notifyIfaceAddrs(ifIndex)
+		}
+	}
+}
+
+func (m *InterfaceMonitor) notifyIfaceAddrs(ifIndex int) {
+	if m.ifaceAddrs[ifIndex] != nil {
+		m.AddrCallback(m.ifaceName[ifIndex], m.ifaceAddrs[ifIndex].Copy())
+	} else {
+		m.AddrCallback(m.ifaceName[ifIndex], m.ifaceAddrs[ifIndex])
+	}
 }
 
-func (m *InterfaceMonitor) storeUpdateAndNotifyOnChange(ifaceExists bool, attrs *netlink.LinkAttrs) {
+func (m *InterfaceMonitor) storeAndNotifyLink(ifaceExists bool, link netlink.Link) {
+	// Store or remove mapping between this interface's index and name.
+	attrs := link.Attrs()
+	ifIndex := attrs.Index
+	ifaceName := attrs.Name
+	if ifaceExists {
+		m.ifaceName[ifIndex] = ifaceName
+	} else {
+		delete(m.ifaceName, ifIndex)
+	}
 	// We need the operstate of the interface; this is carried in the IFF_RUNNING flag.
 	// The IFF_UP flag contains the admin state, which doesn't tell us whether we can
 	// program routes etc.
 	rawFlags := attrs.RawFlags
 	ifaceIsUp := ifaceExists && rawFlags&syscall.IFF_RUNNING != 0
-	ifaceName := attrs.Name
 	ifaceWasUp := m.upIfaces.Contains(ifaceName)
 	logCxt := log.WithField("ifaceName", ifaceName)
 	if ifaceIsUp && !ifaceWasUp {
 		logCxt.Debug("Interface now up")
 		m.upIfaces.Add(ifaceName)
 		m.Callback(ifaceName, StateUp)
+
+		// Get addresses for the link and store and notify those too; then we
+		// don't have to worry about a possible race between the link and address
+		// update channels.
+		new_addrs := set.New()
+		for _, family := range [2]int{netlink.FAMILY_V4, netlink.FAMILY_V6} {
+			addrs, err := netlink.AddrList(link, family)
+			if err != nil {
+				log.WithError(err).Warn("Netlink addr list operation failed.")
+			}
+			for _, addr := range addrs {
+				new_addrs.Add(addr.IPNet.IP.String())
+			}
+		}
+		m.ifaceAddrs[ifIndex] = new_addrs
+		m.notifyIfaceAddrs(ifIndex)
 	} else if ifaceWasUp && !ifaceIsUp {
 		logCxt.Debug("Interface now down")
+
+		// Notify to address callback consumers.
+		delete(m.ifaceAddrs, ifIndex)
```

Comment:
Just occurred to me: I'm not sure this is quite right.  An interface loses its routes when it goes down but not its local addresses.  I don't think we should break the link between IPs and interfaces just because the interface is down (and ideally, we want that link to be established before it comes up to avoid a race where we don't secure the interface until it comes up).